tensorflow/core/common_runtime/eager/eager_op_rewrite_registry.cc

```diff
@@ -22,26 +22,39 @@ EagerOpRewriteRegistry* EagerOpRewriteRegistry::Global() {
   return global_rewrite_registry;
 }
 
-void EagerOpRewriteRegistry::Register(Phase phase,
+void EagerOpRewriteRegistry::Register(Phase phase, int32 ordinal,
                                       std::unique_ptr<EagerOpRewrite> pass) {
-  if (!rewrites_[phase]) {
-    rewrites_[phase] = std::move(pass);
-  } else {
-    TF_CHECK_OK(errors::AlreadyExists(pass->GetDebugInfo().name,
-                                      " is already registered as"
-                                      " EagerOpRewrite for this phase in ",
-                                      pass->GetDebugInfo().file, ":",
-                                      pass->GetDebugInfo().line));
+  auto it_rewrites = rewrites_[phase].cbegin();
+  for (; it_rewrites != rewrites_[phase].cend(); ++it_rewrites) {
+    if (it_rewrites->second == ordinal) {
+      TF_CHECK_OK(errors::AlreadyExists(
+          "Attempting to register Eager Rewriter ", pass->GetDebugInfo().name,
+          " for phase ", phase, " using ordinal ", ordinal,
+          " already occupied by Rewriter ",
+          it_rewrites->first->GetDebugInfo().name));
+    }
+    if (it_rewrites->second > ordinal) {
+      break;
+    }
   }
+  rewrites_[phase].emplace(it_rewrites,
+                           std::make_pair(std::move(pass), ordinal));
 }
 
 Status EagerOpRewriteRegistry::RunRewrite(
     Phase phase, EagerOperation* orig_op,
-    std::unique_ptr<tensorflow::EagerOperation>* out_op) {
-  auto& rewrite = rewrites_[phase];
-  if (rewrite) {
-    TF_RETURN_IF_ERROR(rewrite->Run(orig_op, out_op));
+    std::unique_ptr<EagerOperation>* out_op) {
+  EagerOperation* pre_op = orig_op;
+  std::unique_ptr<EagerOperation>* post_op = out_op;
+  for (auto it_rewrites = rewrites_[phase].cbegin();
+       it_rewrites != rewrites_[phase].cend(); ++it_rewrites) {
+    TF_RETURN_IF_ERROR(it_rewrites->first->Run(pre_op, post_op));
+    if (*post_op != nullptr) {
+      pre_op = post_op->release();
```

Comment:
Do we need to `Release` the intermediate `EagerOperation`s?

cc: @allenlavoie

Reply:
Thanks for the comment. `post_op` is of type `std::unique_ptr<EagerOperation>*`, and therefore `*post_op` is a `std::unique_ptr` which should release its target in the assignment to another pointer. Does this answer your question?

Reply:
`out_op` and `post_op` are the same thing, right? Maybe the main thing to do here is simplify a bit, removing `post_op` and just setting `pre_op = out_op->get()` instead of release/reset.